arangod/GeneralServer/HttpCommTask.cpp

```diff
@@ -293,6 +293,35 @@ bool HttpCommTask<T>::readCallback(asio_ns::error_code ec) {
   return err == HPE_OK && !ec;
 }
 
+template <SocketType T>
+void HttpCommTask<T>::setIOTimeout() {
+  double secs = GeneralServerFeature::keepAliveTimeout();
+  if (secs <= 0) {
+    return;
+  }
+  
+  const bool wasReading = this->_reading;
+  const bool wasWriting = this->_writing;
+  TRI_ASSERT((wasReading && !wasWriting) || (!wasReading && wasWriting));
+  
+  auto millis = std::chrono::milliseconds(static_cast<int64_t>(secs * 1000));
+  this->_protocol->timer.expires_after(millis);
+  this->_protocol->timer.async_wait([=, self = CommTask::weak_from_this()](asio_ns::error_code const& ec) {
+    std::shared_ptr<CommTask> s;
+    if (ec || !(s = self.lock())) {  // was canceled / deallocated
+      return;
+    }
+    
+    auto& me = static_cast<HttpCommTask<T>&>(*s);
+    if ((wasReading && me._reading) ||
```

Comment:
As I understand, this code will always call `asyncReadNextRequest` and thus `_reading` almost always true.

Now, previously I added the `_requestCount` which is now gone again. The reason was the following: imagine the a new request coming in and the timeout expires at the same time. Now both completion handlers are on the queue, first the read completion runs and decides to execute the request, and then restart reading. Now the timeout completion runs, sees that it is still reading and closes the request. Thus we have the same bug as before.

Reply:
I think you are mixing these, this is the http 1.1 version there is no `asyncReadNextRequest` in this class

Reply:
_requestCount was also never increased here

Reply:
VST / HTTP2 now have their own `setIOTimeout()` implementation

Reply:
_reading is not true unless we are onto the next request actually

Reply:
Got it. I thouht we start reading the next request as soon as we called `processRequest`. However, that is indeed not the case here.